### PR TITLE
#162414163 social login authentication

### PR DESCRIPTION
**Pull request by iAfolayan**
What does this PR do?
This PR enable a user to login via Google, Facebook, Twitter, and Linkedin

Description of Task to be completed?
It ensures that a user login through Google, Facebook, Twitter and Linkedin

How should this be manually tested?
- Pull the current repository
- Change to the directory
- Switch to ft-social-login-#162414163 branch
- Run npm install to install all dependencies.

Hit the following endpoints on your browser
-  localhost:3000/api/v1/auth/facebook (for Facebook API)
- localhost:3000/api/v1/auth/google (for Google API)
- localhost:3000/api/v1/auth/twitter (for Twitter API)
- localhost:3000/api/v1/auth/linkedIn (for Linkedin API)

What are the relevant pivotal tracker stories?
#162414163

Screenshots
--- Facebook screenshot
screen shot 2019-01-17 at 11 00 04 am

--- Twitter screenshot
screen shot 2019-01-17 at 11 05 34 am

--- Linkedin screenshot
screen shot 2019-01-17 at 11 08 24 am

--- Google screenshot
screen shot 2019-01-17 at 11 06 21 am

### CODE REVIEW

**Review by Maxfurry — DISMISSED** [review]

File: lib/utils/mockStrategy.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+   * @memberof Strategy
+   */
+  authenticate() {
+    this.callback(null, null, this.newUser, (error, profile) => {
+      this.success(profile);
+    });
+  }
+}
+
+const MockStrategy = Strategy;
```
Comment (codrex) [inline lib/utils/mockStrategy.js]: `const MockStrategy = Strategy;` Please explain to me what is going on here

File: lib/utils/mockStrategy.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    this.callback = strategyCallback;
+  }
+
+  /**
+   * @override
+   * @returns {Function} - Callback function
+   * @memberof Strategy
+   */
+  authenticate() {
+    this.callback(null, null, this.newUser, (error, profile) => {
```
Comment (codrex) [inline lib/utils/mockStrategy.js]: Please explain to me what is going on here

File: test/socialAuth/socialLoginControllers.test.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    };
+
+    const res = {
+      status() {},
+      send() {},
+    };
+    sinon.stub(res, 'status').returnsThis();
+    await socialMediaController.getUserToken(req, res);
+    expect(res.status).to.have.been.calledWith(201);
+  });
```
Comment (codrex) [inline test/socialAuth/socialLoginControllers.test.js:27]: I expected this test to have a happy and a sad path

File: controllers/socialLoginControllers.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+   *
+   * @memberof SocialMediaController
+   */
+  static async socailLoginsCallback(accessToken, refreshToken, profile, done) {
+    const {
+      displayName,
+      emails,
+    } = profile;
+    const name = displayName.split(' ');
+    await User.findOrCreate({
```
Comment (codrex) [inline controllers/socialLoginControllers.js]: ` await User.findOrCreate` should be in the user model manager

File: controllers/socialLoginControllers.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        password: generator.generate({
+          length: 15,
+          uppercase: true,
+          numbers: true,
+          symbols: true,
+          exclude: [''],
+          excludeSimilarCharacters: true
+        }),
+        firstname: name.givenName,
+        lastname: name.familyName,
```
Comment (codrex) [inline controllers/socialLoginControllers.js]: `givenName` `familyName` please I would love to know what inspired these variable names.

File: controllers/socialLoginControllers.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      expiresIn: '24h'
+    };
+    const token = await generateUserToken({ id, email }, time);
+    const userDetails = Object.assign({}, req.user, {
+      msg: created ? 'Welcome!' : 'Welcome back',
+      token
+    });
+    const status = created ? 201 : 200;
+    res.status(status).send(userDetails);
+    // res.redirect(`/?token=${token}`);
```
Comment (codrex) [inline controllers/socialLoginControllers.js]: Remove dead code

File: test/socialAuth/index.test.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  chaiHttp = require('chai-http'),
+  nock = require('nock'),
+  app = require('../../');
+
+const { expect } = chai;
+chai.use(chaiHttp);
+
+describe('Signs up user with their Google account', () => {
+  before(() => {
+    nock('https://accounts.google.com')
```
Comment (codrex) [inline test/socialAuth/index.test.js:12]: `nock('https://accounts.google.com')` should be called within the respective `it blocks`

File: test/socialAuth/socialLoginControllers.test.js (hunk 3 above)
Comment (iAfolayan) [reply test/socialAuth/socialLoginControllers.test.js:27]: Fixed

File: controllers/socialLoginControllers.js (hunk 6 above)
Comment (iAfolayan) [reply controllers/socialLoginControllers.js]: fixed

File: test/socialAuth/index.test.js (hunk 7 above)
Comment (iAfolayan) [reply test/socialAuth/index.test.js:12]: fixed

File: controllers/socialLoginControllers.js (hunk 5 above)
Comment (iAfolayan) [reply controllers/socialLoginControllers.js]: givenName and familyName are part of the response from passportjs but have been fixed.

File: controllers/socialLoginControllers.js (hunk 4 above)
Comment (iAfolayan) [reply controllers/socialLoginControllers.js]: Fixed

**Pull request merged**